decisions/backlog/0023-error-responses.md

```diff
@@ -0,0 +1,83 @@
+* **RFC6749 (https://www.rfc-editor.org/rfc/rfc6749) Section 4.4:** *If the request failed client authentication or is invalid, the authorization server returns an error response as described in Section 5.2.*
+  * **Section 5.2:** *"The authorization server responds with an HTTP 400 (Bad Request) status code (unless specified otherwise) and includes the following parameters with the response:"*
+    *  "error (REQUIRED)"
+    *  "error_description (OPTIONAL)"
+    *  "error_uri (OPTIONAL)"
+
+For example:
+```javascript
+     HTTP/1.1 400 Bad Request
+     Content-Type: application/json;charset=UTF-8
+     Cache-Control: no-store
+     Pragma: no-cache
+
+     {
+       "error":"..."
+     }   
+```
+
+#### Error Responses
+* **Tech. Specs. (both V1 and V2), Section 6.5/6.9:**
+  * *"Whenever a host system returns an error response, it MUST send a HTTP response such that:"*
+    * *"with response body the error response"*
+  * *"A error response is a JSON object with the following properties:*
+    * *"code: a error response code encoded as a String"*
+    * *"message: a error message encoded as a String"*
+
+For example:
+```javascript
+     HTTP/1.1 400 Bad Request
+     Content-Type: application/json;charset=UTF-8
+     Cache-Control: no-store
+     Pragma: no-cache
+
+     {
+       "code":"...",
+       "message":"..."
+     }   
+```
+
+As shown, these requirements conflict when it comes to the structure of the body of the error response. Therefore, the Tech. Specs. need an update to eliminate this conflict.
+
+## Summary
+
+In general, we propose to stick to standards as much as possible. Hence, we propose to adjust the specifications in the section `Error Responses` (6.5/6.9) so that they do not conflict with a standard. In this case, that means that the specification for the error response can not apply to the `Action Authenticate`. 
+
+
+## Decision
+
+### Technical Specification
+
+**6.9. Error Responses**
+
+Whenever a host system returns an error response for any Action **other** than the `Action Authenticate`, it MUST send a HTTP response such that
+* the HTTP Status Code equals the HTTP Status Code defined for the respective error response code (see Error Codes Table)
+* with content type set to application/json, and
+* with response body the error response
+
+A error response is a JSON object with the following properties:
+* code: a error response code encoded as a String
+* message: a error message encoded as a String
+
+A error response code is a value from column Error Response Code from the table below.
+
+A error message is a human-readable error description. Example values are in column Example Message in the table below.
+
+## Consequences
+
+1. As this ADR is more a correction than a change, the Tech Specs Version number can be updated from 2.0.x to 2.1.y.
+2. This correction results in having multiple structures for `error responses`. 
```

Comment:
I understand that the auth path is common across versions v1.x.x, v2.0.0, v2.0.1, v2.1.0. If you incorporate ADR0024 into the system and perform a version update, all versions will have error responses following the v2.1.0 specification, making it impossible to fulfill the requirements of other specifications (e.g. v2.0.0). Is this understanding correct?

Alternatively, is it permissible for a system implementing an existing specification (e.g., v1.0.1) to return responses according to this ADR? Is it acceptable for the authentication endpoint for calling ListFootprint v1.0.1 to be v2.1.0?

Reply:
@nri-kosuke-nakamura-01  - Will you be able to join our Session tomorrow (18/10/2023 @ 4pm CEST)? Can you please share your email, so I can invite you as well? Thanks!

Reply:
Due to personal circumstances, I could only participate in the final 10 minutes of the meeting. During that time, the discussion in the meeting revolved around what issues?
Did you convey anything to NRI uemura-san regarding this issue?

---- 

In practice, software implementers rarely use the detailed error messages returned from other authentication endpoints for fine-grained control. They can typically develop software using only the information in the HTTP headers, including the HTTP error response status of other authentication endpoints. Therefore, although this ADR is important, there is no need to be overly concerned about it.

As an extension or superset to the technical specification, it would suffice to establish a convention where each software indicates a compliant version of the authentication endpoint.

While it would be technically correct to support multiple authentication endpoints or switch the response specifications based on the version, this may not be well-received by many users.